### PR TITLE
[FEATURE] Add time span hooks, tests and docs

### PR DESCRIPTION
**Pull request by mk-mxp**
The first hook to utilize the modified values hook.

I called the interface `TimeSpan` in accordance to your `AbstractTimeSpan` class. But I would also call it `DateTimeSpan` to better reflect that is used for both.

### CODE REVIEW

**Comment by oliverklee** [comment]
> I called the interface `TimeSpan` in accordance to your `AbstractTimeSpan` class. But I would also call it `DateTimeSpan` to better reflect that is used for both.

Please feel free to name the new interface the way you find most fitting.

File: Classes/Hooks/Interfaces/TimeSpan.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * Use this interface for hooks concerning the date span.
+ *
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+interface TimeSpan extends Hook
+{
+    /**
+     * Modifies the date span string.
+     *
+     * The date format for the date parts are configured in TypoScript (`dateFormatYMD` etc).
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/TimeSpan.php]: Please add the missing period directly after "etc".

File: Classes/Hooks/Interfaces/TimeSpan.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+     * This allows modifying the assembly of start and end date to the date span.
+     * E.g. for Hungarian: '01.-03.01.2019' -> '2019.01.01.-03.'.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/TimeSpan.php]: Please add the missing comma after "E.g.".

Comment (oliverklee) [inline Classes/Hooks/Interfaces/TimeSpan.php]: This should not be left to be fixed in a hook. Instead, I'd prefer we fix the original code instead and move the way shortened dates are assembled to the locallang files.

**Comment by mk-mxp** [comment]
> Please let me know whether this hook still is needed if we fix the date assembly in the extension itself.

"Fixing" could mean some more in-depth investigation into I18N of seminars. One main principle of I18N is not to assemble output from re-usable parts (which requires in-depth knowledge of grammar of all possible languages and thus is a bad thing). Such a deep fix requires a complete re-thinking of frontend output in general.

The hook allows all options from date() and then implement the required way to assemble it. If I have to do this with variable replacement, I need variables for all possible combinations of dates. Looking at my Hungarian example you can see, that the begin date is always in full length, but the end date becomes shortened for a short date format.

Looking at what I would need only, it is OK to move the assembly to translation files if there is variable replacement for 2 dates / times in one string with the same flexibility as date() has itself (12 hours am/pm vs. 24h format, dates with abbreviated monthes etc.).

There is an existing flaw that might be thought of for such a change, too: Having different begin and end dates a time span output like "03.10. 10:00h - 05.10. 17:00h" should be possible.

These are my thoughts, it's your decision. I18N is a hard thing, and all concepts in real live have issues.

**Comment by mk-mxp** [comment]
May I continue here? Or are You working on #470 and this is obsolete?

**Comment by oliverklee** [comment]
I'll review this here, and #470 is on hold. Please wait with new changes until I've finished my review here. Thanks!

File: Classes/Hooks/Interfaces/TimeSpan.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OliverKlee\Seminars\Hooks\Interfaces;
+
+/**
+ * Use this interface for hooks concerning the date span.
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/TimeSpan.php]: This should be consistent with the class name (or the comment should explain the difference).

File: Classes/Hooks/Interfaces/TimeSpan.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+     * E.g. for Hungarian: '01.-03.01.2019' -> '2019.01.01.-03.'.
+     *
+     * @param string $dateSpan the date span produced by `AbstractTimeSpan::getDate()`
+     * @param string $beginDate the formatted begin date part (`dateFormatYMD`)
+     * @param string $dash the glue used by `AbstractTimeSpan::getDate()` (may be HTML encoded)
+     * @param string $endDate the formatted end date part (`dateFormatYMD`)
+     *
+     * @return string the modified date span to use
+     */
+    public function modifyDateSpan(string $dateSpan, string $beginDate, string $dash, string $endDate): string;
```
Comment (oliverklee) [inline Classes/Hooks/Interfaces/TimeSpan.php]: Wouldn't it be more flexible to pass the begin and end date as timestamps?

File: Classes/OldModel/AbstractTimeSpan.php [hunk 5]
```diff
@@ -305,4 +327,21 @@ public function getNumberOfPlaces(): int
      * @return string our places or an empty string if the timespan has no places
      */
     abstract public function getPlaceShort(): string;
+
+    /**
+     * Gets the hook provider for the list view.
```
Comment (oliverklee) [inline Classes/OldModel/AbstractTimeSpan.php]: "for the list view" looks like a copy'n'paste error.

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php [hunk 6]
```diff
@@ -22,6 +24,31 @@ final class AbstractTimeSpanTest extends FunctionalTestCase
...
+
+    /**
+     * @var string
+     */
+    const DATE_FORMAT_M = '%m';
+
+    /**
+     * @var string
+     */
+    const DATE_FORMAT_Y = '%Y';
```
Comment (oliverklee) [inline Tests/Functional/OldModel/AbstractTimeSpanTest.php]: Nice!

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php [hunk 7]
```diff
@@ -111,4 +146,260 @@ public function getTimeForBeginTimeAndEndTimeOnDifferentDaysReturnsBothTimesWith
...
+     */
+    public function getTimeForNoTimeNotCallsHook()
+    {
+        $hook = $this->createMock(TimeSpan::class);
+        $hook->expects(self::never())->method('modifyDateSpan');
+        $hook->expects(self::never())->method('modifyTimeSpan');
+
+        $hookClass = \get_class($hook);
+        $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][TimeSpan::class][] = $hookClass;
+        GeneralUtility::addInstance($hookClass, $hook);
```
Comment (oliverklee) [inline Tests/Functional/OldModel/AbstractTimeSpanTest.php]: I think we also need the `getInstance` calls in `tearDown` to get rid of the instances.

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php [hunk 8]
```diff
@@ -111,4 +146,260 @@ public function getTimeForBeginTimeAndEndTimeOnDifferentDaysReturnsBothTimesWith
...
+
+        $this->subject->getTime();
+    }
+
+    /**
+     * @test
+     */
+    public function getTimeForBeginTimeAndEndTimeOnDifferentTimesCallsHookAndReturnsModifiedValue()
```
Comment (oliverklee) [inline Tests/Functional/OldModel/AbstractTimeSpanTest.php]: Returns -> Uses

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php [hunk 9]
```diff
@@ -111,4 +146,260 @@ public function getTimeForBeginTimeAndEndTimeOnDifferentDaysReturnsBothTimesWith
...
+    {
+        $this->subject->setBeginDateAndTime(\mktime(9, 50, 0, 1, 1, 2010));
+        $this->subject->setEndDateAndTime(\mktime(18, 30, 0, 1, 1, 2010));
+
+        $hook = $this->createMock(TimeSpan::class);
+        $hook->expects(self::never())->method('modifyDateSpan');
+        $hook->expects(self::once())->method('modifyTimeSpan')
+            ->with('09:50&#8211;18:30', '09:50', '&#8211;', '18:30')
+            ->willReturn('modified');
```
Comment (oliverklee) [inline Tests/Functional/OldModel/AbstractTimeSpanTest.php]: I propose putting `modified` in a variable to make the connection more obvious. Ditto in the other respective test.

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php [hunk 10]
```diff
@@ -111,4 +146,260 @@ public function getTimeForBeginTimeAndEndTimeOnDifferentDaysReturnsBothTimesWith
...
+        $hook->expects(self::once())->method('modifyDateSpan')
+            ->with('01.&#8211;03.01.2010', '01.01.2010', '&#8211;', '03.01.2010')
+            ->willReturn('modified');
+        $hook->expects(self::never())->method('modifyTimeSpan');
+
+        $hookClass = \get_class($hook);
+        $GLOBALS['TYPO3_CONF_VARS']['EXTCONF']['seminars'][TimeSpan::class][] = $hookClass;
+        GeneralUtility::addInstance($hookClass, $hook);
+
+        self::assertSame('modified', $this->subject->getDate());
```
Comment (oliverklee) [inline Tests/Functional/OldModel/AbstractTimeSpanTest.php]: I like the tests a lot.

File: Classes/Hooks/Interfaces/TimeSpan.php (hunk 4 above)
Comment (mk-mxp) [reply Classes/Hooks/Interfaces/TimeSpan.php]: It'd need to get the configured format strings into the hooked-in method, also. That's why I chose to use the pre-formatted strings. I'll take a look into a way of passing all that without dozens of parameters.

File: Classes/OldModel/AbstractTimeSpan.php (hunk 5 above)
Comment (mk-mxp) [reply Classes/OldModel/AbstractTimeSpan.php]: Indeed, it is.

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php (hunk 7 above)
Comment (mk-mxp) [reply Tests/Functional/OldModel/AbstractTimeSpanTest.php]: IIRC, no. Nimut and the Typo3 testing framework do a reset of the instance buffer (and a lot more to start with a fresh Typo3 instance) before each test. Its all in `setUp()` and the main reason why it takes endless amounts of time to run the tests (and why there is a DB copy left after the tests end).

**Comment by mk-mxp** [comment]
FYI: It will take some time to provide the changes.

File: Tests/Functional/OldModel/AbstractTimeSpanTest.php (hunk 7 above)
Comment (mk-mxp) [reply Tests/Functional/OldModel/AbstractTimeSpanTest.php]: Just tested: This is not required.

File: Classes/Hooks/Interfaces/TimeSpan.php (hunk 4 above)
Comment (mk-mxp) [reply Classes/Hooks/Interfaces/TimeSpan.php]: I could pass `$this` (the instance of `\Tx_Seminars_OldModel_AbstractTimeSpan`) to the methods. One could use `::getBeginDateAsTimestamp()` as well as `getConfValueString('dateFormatYMD')` to get the required values from the caller. I would keep `$dateSpan` (for smaller adjustments) and `$dash` (to indicated HTML encoded or not) also. OK?

Comment (oliverklee) [reply Classes/Hooks/Interfaces/TimeSpan.php]: @mk-mxp Sounds good.

**Comment by oliverklee** [comment]
@mk-mxp Thanks for the update of this PR. I have some other things on my plate (including some bug fixes for seminars) and then will take care of reviewing this PR afterwards. It't not forgotten!

**Pull request merged**